4-custom-evaluations.ipynb

```diff
@@ -0,0 +1,744 @@
+    "    return response, LLMObs.export_span()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our RAG workflow is ready! Try a question about LLM Observability. What do you think of the answer quality?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Answer: Use the steps provided below to run a simple Python script that generates an LLM Observability trace. To begin, ensure you have the necessary prerequisites in place, such as a Datadog API key and an OpenAI API key stored in your environment as `OPENAI_API_KEY`. You will also need to have the OpenAI Python library installed. Once you have these prerequisites set up, you can proceed with running the script as outlined in the quickstart guide.\n",
+      "Context: ['Prerequisites\\n\\n- LLM Observability requires a Datadog API key (see [the instructions for creating an API key][7]).\\n- The example script below uses OpenAI, but you can modify it to use a different provider. To run the script as written, you need:\\n    - An OpenAI API key stored in your environment as `OPENAI_API_KEY`. To create one, see [Account Setup][4] and [Set up your API key][6] in the OpenAI documentation.\\n    - The OpenAI Python library installed. See [Setting up Python][5] in the OpenAI documentation for instructions.', 'Command-line quickstart\\n\\nUse the steps below to run a simple Python script that generates an LLM Observability trace.']\n"
+     ]
+    }
+   ],
+   "source": [
+    "STARTER_QUESTION = \"How do I get started?\"\n",
+    "\n",
+    "answer, _ = ask_docs(STARTER_QUESTION, retriever=base_retriever)\n",
+    "\n",
+    "print(\"Answer: {}\".format(answer))\n",
+    "print(\"Context: {}\".format([reference.text for reference in answer.source_nodes]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Recursive Retriever"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's implement a recursive retriever and plug that into our `ask_docs` workflow. \n",
+    "\n",
+    "A recursive retriever first builds a graph of small chunks that have references to larger parent chunks. At query-time, smaller chunks are retrieved first, and then we follow references to bigger chunks. This enhances the context we pass to the augmented generation step. For more information on recursive retrieval, see Llamaindex's [recursive retrieval](https://docs.llamaindex.ai/en/stable/examples/retrievers/recursive_retriever_nodes/) guide.\n",
+    "\n",
+    "Since our raw documents are in markdown, there's already an implicit parent-child relationship between different text chunks. Llamaindex provides helpful utility functions to automatically parse these relationships and form an index that is searchable using their `RecursiveRetriever` module"
```

Comment:
```suggestion
    "Since our raw documents are in Markdown, there's already an implicit parent-child relationship between different text chunks. LlamaIndex provides helpful utility functions to automatically parse these relationships and form an index that is searchable using their `RecursiveRetriever` module."
```